### PR TITLE
 Breaking Change: Use dynamo db

### PR DESCRIPTION
**Pull request by wjarvis1**

### CODE REVIEW

File: lib/compat.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
+const through = require('through2');
```
Comment (cfellin1) [inline lib/compat.js]: How is this different from the lib/compat in dynastar?
https://github.secureserver.net/opensource/dynastar/blob/master/index.js

File: lib/build-head.js [hunk 2]
```diff
@@ -1,51 +1,48 @@
...
 
-  var BuildHead = datastar.define('build_head', {
-    schema: datastar.schema.object({
-      env: cql.text(),
-      name: cql.text(),
-      build_id: cql.text(),
-      previous_build_id: cql.text(),
-      rollback_build_ids: cql.map(cql.text(), cql.text()), // timestamp is stored as string
-      create_date: cql.timestamp({ default: 'create' }),
-      udpate_date: cql.timestamp({ default: 'update' }),
```
Comment (msluther) [inline lib/build-head.js]: bye bye udpate. 😂

File: lib/compat.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+const ls = require('list-stream');
+
+/**
+ * Datastar API compatibility
+ * @class
+ */
+class Compat {
```
Comment (msluther) [inline lib/compat.js]: You don't need this anymore, right?

File: package.json [hunk 4]
```diff
@@ -1,12 +1,13 @@
 {
-  "name": "warehouse-models",
-  "version": "5.9.1",
+  "name": "warehouse-models-dynamo",
+  "version": "6.0.0-0",
```
Comment (cfellin1) [inline package.json]: Where did the '-0' come from? Also, shouldn't we be doing the version bump in a separate commit?

Comment (msluther) [inline package.json]: why the `-0`?

File: test/integration.test.js [hunk 5]
```diff
@@ -37,16 +45,11 @@ describe('registry-data (integration)', function () {
...
-    datastar.connect(done);
-  });
-
-  after(function (done) {
-    datastar.close(done);
+    liveness.waitForServices({
+      clients: [dynamoDriver],
+      waitSeconds: 60
+    }).then(() => done())
+      .catch(done);
```
Comment (msluther) [inline test/integration.test.js]: the before/after mocha functions can be async as well, so you could use `await` here if you wanted. otherwise this is fine.

File: test/integration.test.js [hunk 6]
```diff
@@ -264,23 +251,15 @@ describe('registry-data (integration)', function () {
...
+        done();
       });
     });
 
     it('returns a falsey value for an unknown id', function (done) {
       BuildHead.findOne({
         name: 'email',
-        env: 'proddd'
+        env: 'proddd',
+        locale: 'en-US'
```
Comment (msluther) [inline test/integration.test.js]: is that required now? was there a default before? why do we need locale here?

File: test/unit.test.js [hunk 7]
```diff
@@ -1,71 +1,30 @@
...
-
-  it('should have a method to fetch the binary blob of versioned content', function () {
-    assume(dal.Version.prototype.getAttachment).is.a('function');
-    assume(dal.Version.prototype.getAttachment.length).to.equal(2);
-  });
-
-  it('should have a method to fetch and create a build payload from versioned content', function () {
-    assume(dal.Version.prototype.forBuild).is.a('function');
-    assume(dal.Version.prototype.forBuild.length).to.equal(2);
-  });
```
Comment (msluther) [inline test/unit.test.js]: None of these functions are relevant anymore?

File: lib/compat.js (hunk 3 above)
Comment (wjarvis1) [reply lib/compat.js]: @msluther oops, forgot to remove this file. it isn't used anymore

File: test/integration.test.js (hunk 6 above)
Comment (wjarvis1) [reply test/integration.test.js]: @msluther because we are calling findOne, we need to provide locale since it is used as the range key after switching to dynamo (not particular to `locale` but just whatever is being used as the range key).  was getting an error about number of conditions on the keys is invalid if this range key is missing

Comment (msluther) [reply test/integration.test.js]: Do we have any callers today that don't provide a locale? I'm okay with it being required, we just need to make sure we catch callers and make sure they're providing it.

File: package.json (hunk 4 above)
Comment (wjarvis1) [reply package.json]: @jcrugzz  I thought we were supposed to add `-0` to indicate it as a pre-release for some reason?

File: test/integration.test.js (hunk 6 above)
Comment (wjarvis1) [reply test/integration.test.js]: @msluther I am not sure. I think we are migrating all the repos that we own which consumes this model so we can make sure that these provide whatever the range key that might not be previously provided? Not sure if we already have a plan for repos that we do not own that use warehouse-models, but since this is going to be published under a new major, maybe the callers can just update the findOne calls themselves when they upgrade? I can add a blurb to the README about this  (cc: @jcrugzz)

Comment (jcrugzz) [reply test/integration.test.js]: @msluther we always provide locale when doing a findOne. Cassandra was just more flexible in that a findOne without a locale did a findAll and then just returned the first value

File: package.json (hunk 4 above)
Comment (jcrugzz) [reply package.json]: @wjarvis1 @cfellin1 In the other models I did this, but im considering we just publish this as a new major and leverage the npm aliasing method to install 2 versions of the same package for migration purposes. and yes we should do versioning after its merged.

File: lib/build-head.js [hunk 8]
```diff
@@ -1,51 +1,48 @@
...
+    hashKey: 'key',
+    rangeKey,
+    timestamps: true, // adds updatedAt and createdAt
+    tableName: 'WarehouseBuildHead',
+    schema: {
+      key: Joi.string(),
+      env: Joi.string(),
+      name: Joi.string(),
+      buildId: Joi.string(),
+      previousBuildId: Joi.string(),
```
Comment (jcrugzz) [inline lib/build-head.js]: we might want to allow(null) here in case it appears when migrating but not sure it makes a difference
```suggestion
      previousBuildId: Joi.string().allow(null),
```

File: lib/package.js [hunk 9]
```diff
@@ -3,123 +3,73 @@
...
-        // Save these statements for when the Package statements get executed so
-        // we can merge together. This should ensure the proper object reference
-        // is set
-        //
-        options.statements = statements;
-        options.statements.strategy = 'batch';
-        next();
-      });
+  ['one', 'all'].forEach(function (type) {
+    model.after(['find', type].join(':'), function (result, callback) {
```
Comment (jcrugzz) [inline lib/package.js]: this stuff no longer works, we can't auto deserialize this object magically.

File: lib/package.js [hunk 10]
```diff
@@ -180,13 +130,13 @@ module.exports = function pkg(datastar, models) {
...
+  }
+
 
-  //
   // Deserialize some of our custom data structures stored as text until we know
   // the schema will enforce these for us nicely
-  //
-  Pkg.deserialize = function deserialize(results) {
+
+  function deserialize(results) {
```
Comment (jcrugzz) [inline lib/package.js]: add this as an extra function passed to `Dynastar` but modify it to only support accepting a single object. we will need to integrate this in places where we expected the deserializing to happen previously in the application

File: lib/release-line-dep.js [hunk 11]
```diff
@@ -1,30 +1,35 @@
...
-      }
+module.exports = function releaseLineDep(dynamo) {
+  const createKey = (data) => {
+    return [data.pkg, data.version].join('!');
+  };
+  const rangeKey = 'dependent';
+  const model = dynamo.define('ReleaseLineDep', {
+    hashKey: 'key',
+    rangeKey,
+    tableName: 'WarehouseReleaseLineDep',
```
Comment (jcrugzz) [inline lib/release-line-dep.js]:
```suggestion
    tableName: 'WrhsReleaseLineDep',
```

File: lib/release-line-head.js [hunk 12]
```diff
@@ -1,27 +1,27 @@
...
-    }).partitionKey('pkg'),
-    with: {
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+module.exports = function releaseLineHead(dynamo) {
+  const hashKey = 'pkg';
+  const model = dynamo.define('ReleaseLineHead', {
+    hashKey,
+    tableName: 'WarehouseReleaseLineHead',
```
Comment (jcrugzz) [inline lib/release-line-head.js]:
```suggestion
    tableName: 'WrhsReleaseLineHead',
```

File: lib/release-line-head.js [hunk 13]
```diff
@@ -1,27 +1,27 @@
...
+    schema: {
+      pkg: Joi.string(),
+      previousVersion: Joi.string(),
```
Comment (jcrugzz) [inline lib/release-line-head.js]:
```suggestion
      previousVersion: Joi.string().allow(null),
```

File: lib/release-line-dep.js [hunk 14]
```diff
@@ -1,30 +1,35 @@
...
+    schema: {
+      key: Joi.string(),
+      pkg: Joi.string(),
+      previousVersion: Joi.string(),
```
Comment (jcrugzz) [inline lib/release-line-dep.js]:
```suggestion
      previousVersion: Joi.string().allow(null),
```

File: lib/release-line.js [hunk 15]
```diff
@@ -1,27 +1,31 @@
...
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+module.exports = function releaseLine(dynamo) {
+  const createKey = (data) => {
+    return [data.pkg, data.version].join('!');
+  };
+  const model = dynamo.define('ReleaseLine', {
+    hashKey: 'key',
+    tableName: 'WarehouseReleaseLine',
```
Comment (jcrugzz) [inline lib/release-line.js]:
```suggestion
    tableName: 'WrhsReleaseLine',
```

File: lib/release-line.js [hunk 16]
```diff
@@ -1,27 +1,31 @@
...
+    schema: {
+      key: Joi.string(),
+      pkg: Joi.string(),
+      previousVersion: Joi.string(),
```
Comment (jcrugzz) [inline lib/release-line.js]:
```suggestion
      previousVersion: Joi.string().allow(null),
```

File: test/fixtures/build.json [hunk 17]
```diff
@@ -7,5 +7,5 @@
   "version": "2.1",
   "fingerprints": ["a", "a.gz"],
   "artifacts": ["a/email.js"],
-  "recommended": []
+  "recommended": ["a/email.js"]
```
Comment (jcrugzz) [inline test/fixtures/build.json]: how come these needed to be changed?

File: lib/build-file.js [hunk 18]
```diff
@@ -1,63 +1,40 @@
...
-    with: {
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+module.exports = function buildFile(dynamo) {
+  const hashKey = 'fingerprint';
+  const model = dynamo.define('BuildFile', {
+    hashKey,
+    timestamps: true, // adds updatedAt and createdAt
+    tableName: 'WarehouseBuildFile',
```
Comment (jcrugzz) [inline lib/build-file.js]:
```suggestion
    tableName: 'WrhsBuildFile',
```

File: lib/build-head.js [hunk 19]
```diff
@@ -1,51 +1,48 @@
...
-      .clusteringKey('locale'),
-    with: {
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+  const model = dynamo.define('BuildHead', {
+    hashKey: 'key',
+    rangeKey,
+    timestamps: true, // adds updatedAt and createdAt
+    tableName: 'WarehouseBuildHead',
```
Comment (jcrugzz) [inline lib/build-head.js]:
```suggestion
    tableName: 'WrhsBuildHead',
```

File: lib/build-head.js [hunk 20]
```diff
@@ -1,51 +1,48 @@
...
+      name: Joi.string(),
+      buildId: Joi.string(),
+      previousBuildId: Joi.string(),
+      rollbackBuildIds: Joi.object().allow(null),
+      version: Joi.string(),
+      locale: Joi.string(),
+      cdnUrl: Joi.string(), // URL of CDN to be used as a base for all the artifacts
+      fingerprints: dynamo.types.stringSet(),
+      artifacts: dynamo.types.stringSet(),
+      recommended: dynamo.types.stringSet()
```
Comment (jcrugzz) [inline lib/build-head.js]:
```suggestion
      recommended: dynamo.types.stringSet().allow(null)
```

File: lib/build.js [hunk 21]
```diff
@@ -1,49 +1,47 @@
...
+  const createKey = (data) => {
+    return [data.name, data.env, data.version].join('!');
+  };
+  const rangeKey = 'locale';
+
+  const model = dynamo.define('Build', {
+    hashKey: 'key',
+    rangeKey,
+    timestamps: true, // adds updatedAt and createdAt
+    tableName: 'WarehouseBuild',
```
Comment (jcrugzz) [inline lib/build.js]:
```suggestion
    tableName: 'WrhsBuild',
```

File: lib/dependent-of.js [hunk 22]
```diff
@@ -1,28 +1,28 @@
...
-    }).partitionKey('pkg'),
-    with: {
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+module.exports = function dependentOf(dynamo) {
+  const hashKey = 'pkg';
+  const model = dynamo.define('DependentOf', {
+    hashKey,
+    tableName: 'WarehouseDependentOf',
```
Comment (jcrugzz) [inline lib/dependent-of.js]:
```suggestion
    tableName: 'WrhsDependentOf',
```

File: lib/dependent.js [hunk 23]
```diff
@@ -1,28 +1,27 @@
...
-    }).partitionKey('name'),
-    with: {
-      compaction: {
-        class: 'LeveledCompactionStrategy'
-      }
+module.exports = function dependent(dynamo) {
+  const hashKey = 'name';
+  const model = dynamo.define('Dependent', {
+    hashKey,
+    tableName: 'WarehouseDependent',
```
Comment (jcrugzz) [inline lib/dependent.js]:
```suggestion
    tableName: 'WrhsDependent',
```

File: lib/package.js [hunk 24]
```diff
@@ -3,123 +3,73 @@
...
-    schema: datastar.schema.object({
-      name: cql.text(),
-      version: cql.text(),
-      description: cql.text(),
-      main: cql.text(),
+module.exports = function pkg(dynamo) {
+  const hashKey = 'name';
+  const model = dynamo.define('Package', {
+    hashKey,
+    tableName: 'WarehousePackage',
```
Comment (jcrugzz) [inline lib/package.js]:
```suggestion
    tableName: 'WrhsPackage',
```

File: lib/package.js [hunk 25]
```diff
@@ -3,123 +3,73 @@
...
-  Pkg.extractLatest = function (data) {
-    var version = (data.distTags || data['dist-tags'] || {}).latest;
+  var Pkg = new Dynastar({
+    model,
+    hashKey,
+    // Functions to attach to this class
+    fromPublish
+  });
+
+  // function extractLatest(data) {
```
Comment (jcrugzz) [inline lib/package.js]: We are still using this below

File: lib/package.js [hunk 26]
```diff
@@ -129,9 +79,9 @@ module.exports = function pkg(datastar, models) {
    *
    * @returns {Object} ?
    */
-  Pkg.fromPublish = function (payload) {
-    var replica = assign({}, this.extractLatest(payload));
-    var schemaFields = Pkg.schema.mappedFields();
+  function fromPublish({ pkg, payload }) {
+    var replica = assign({}, pkg.extractLatest(payload));
```
Comment (jcrugzz) [inline lib/package.js]: extractLatest that is commented out is whats used here. And in this case, we actually just need the payload here, this is not a function that was on the prototype of the model, it was a static function attached to the class/constructor.
```suggestion
    var replica = assign({}, extractLatest(payload));
```

File: lib/package.js (hunk 25 above)
Comment (jcrugzz) [inline lib/package.js]:
```suggestion
  function extractLatest(data) {
```

File: lib/package.js [hunk 27]
```diff
@@ -3,123 +3,73 @@
...
-    var version = (data.distTags || data['dist-tags'] || {}).latest;
+  var Pkg = new Dynastar({
+    model,
+    hashKey,
+    // Functions to attach to this class
+    fromPublish
+  });
+
+  // function extractLatest(data) {
+  //   var version = (data.distTags || data['dist-tags'] || {}).latest;
```
Comment (jcrugzz) [inline lib/package.js]:
```suggestion
     var version = (data.distTags || data['dist-tags'] || {}).latest;
```

File: lib/package.js [hunk 28]
```diff
@@ -3,123 +3,73 @@
...
+
+  //   return (data.versions || {})[version] || {};
```
Comment (jcrugzz) [inline lib/package.js]:
```suggestion
     return (data.versions || {})[version] || {};
```

File: lib/package.js [hunk 29]
```diff
@@ -3,123 +3,73 @@
...
+  // }
```
Comment (jcrugzz) [inline lib/package.js]:
```suggestion
   }
```

File: lib/package.js [hunk 30]
```diff
@@ -3,123 +3,73 @@
...
-      name: cql.text(),
-      version: cql.text(),
-      description: cql.text(),
-      main: cql.text(),
+module.exports = function pkg(dynamo) {
+  const hashKey = 'name';
+  const model = dynamo.define('Package', {
+    hashKey,
+    tableName: 'WarehousePackage',
+    schema: {
```
Comment (jcrugzz) [inline lib/package.js]: define this as its own variable above. `const schema = { }` and pass it in as `schema` in this `dynamo.define`, we need to use `Object.keys(schema)` to replace the previous `model.schema.mappedFields` defined below

File: lib/package.js (hunk 29 above)
Comment (jcrugzz) [reply lib/package.js]: spacing might be off on these so it might have to be corrected locally

File: lib/package.js [hunk 31]
```diff
@@ -129,9 +79,9 @@ module.exports = function pkg(datastar, models) {
    *
    * @returns {Object} ?
    */
-  Pkg.fromPublish = function (payload) {
-    var replica = assign({}, this.extractLatest(payload));
-    var schemaFields = Pkg.schema.mappedFields();
+  function fromPublish({ pkg, payload }) {
+    var replica = assign({}, pkg.extractLatest(payload));
+    var schemaFields = model.schema.mappedFields();
```
Comment (jcrugzz) [inline lib/package.js]: Since we can define schema as a separate variable in the scope of the model defining here, we can use it in the function here
```suggestion
    var schemaFields = Object.keys(schema);
```

File: lib/version.js [hunk 32]
```diff
@@ -5,35 +5,44 @@ var url = require('url');
...
-      version_id: cql.text(),
-      name: cql.text(),
-      version: cql.text(),
+module.exports = function version(dynamo) {
+  const hashKey = 'name';
+  const rangeKey = 'version';
+  const model = dynamo.define('Version', {
+    hashKey,
+    rangeKey,
+    tableName: 'WarehouseVersion',
```
Comment (jcrugzz) [inline lib/version.js]:
```suggestion
    tableName: 'WrhsVersion',
```

File: test/unit.test.js (hunk 7 above)
Comment (jcrugzz) [reply test/unit.test.js]: Some are but the prototype ones were moved to static functions that take the object as a parameter rather than relying on being on a model prototype and using `this`

File: test/unit.test.js [hunk 33]
```diff
@@ -1,71 +1,30 @@
...
-    assume(second.source).is.a('string');
-    assume(second.sourcemap).is.a('string');
-  });
-
-  it('should return a dal whos Build and BuildHead prototype has a fetchFiles function', function () {
-    assume(dal.Build.prototype.fetchFiles).is.a('function');
-    assume(dal.BuildHead.prototype.fetchFiles).is.a('function');
-  });
-
-  it('should contain a defined function on the Package model that should strip values', function () {
```
Comment (jcrugzz) [inline test/unit.test.js]: these should still be tested and be on the model as a static function with the new `Dynastar`

File: test/fixtures/build.json (hunk 17 above)
Comment (wjarvis1) [reply test/fixtures/build.json]: if recommended was only passed an empty array, when we test it later in `assume(result.recommended )...`, result.recommended becomes undefined. This happens with artifacts as well, where an empty array past in results in undefined field in the result.

File: lib/package.js [hunk 34]
```diff
@@ -3,123 +3,66 @@
...
+module.exports = function pkg(dynamo) {
+  const hashKey = 'name';
+  const schema = {
+    name: Joi.string(),
+    version: Joi.string(),
+    description: Joi.string(),
+    main: Joi.string(),
+    // Git hash
+    gitHead: Joi.string(),
+    // text for backwards compat but support this as JSON in the future
```
Comment (msluther) [inline lib/package.js]: I don't think this comment is relevant anymore.

File: test/integration.test.js [hunk 35]
```diff
@@ -182,17 +181,8 @@ describe('registry-data (integration)', function () {
...
-        result.fetchFiles(function (error, files) {
-          assume(error).is.falsey();
-          assume(files).is.an('array');
-          files.forEach(function (file) {
-            assume(file).is.an('object');
-            assume(file.url).is.a('string');
-          });
-          done();
-        });
+        assume(result.recommended.sort()).eql(BuildFixture.recommended.sort());
```
Comment (msluther) [inline test/integration.test.js]: does that `.eql` work? these are arrays, so `.eql` by default would be doing a reference equality, right? should this be `.deep.equals`?

File: test/mocha.opts [hunk 36]
```diff
@@ -1,2 +1,2 @@
 --recursive
--t 30000
+-t 60000
```
Comment (msluther) [inline test/mocha.opts]: so tests got slower with localstack? 😢

File: test/integration.test.js (hunk 35 above)
Comment (jcrugzz) [reply test/integration.test.js]: @msluther `equals` is shallow `eql` is shorthand for deep equals

File: test/mocha.opts (hunk 36 above)
Comment (jcrugzz) [reply test/mocha.opts]: @msluther its just for defensiveness so we can start the tests concurrently with starting localstack

File: test/unit.test.js [hunk 37]
```diff
@@ -3,69 +3,55 @@
...
 
   it('should contain a defined function on the Package model that should strip values', function () {
     var secondFixture = clone(packageStrip);
     assume(dal.Package.fromPublish).is.a('function');
     assume(dal.Package.deserialize).is.a('function');
-    assume(dal.Package.fromPublish(packageStrip.before)).to.eql(packageStrip.after);
-    assume(dal.Package.deserialize(secondFixture.after)).to.eql(
-      dal.Package.deserialize(dal.Package.fromPublish(secondFixture.before))
+    assume(dal.Package.fromPublish({ payload: packageStrip.before })).to.eql(packageStrip.after);
+    assume(dal.Package.deserialize({ results: secondFixture.after })).to.eql(
```
Comment (jcrugzz) [inline test/unit.test.js]:
```suggestion
    assume(dal.Package.deserialize(secondFixture.after)).to.eql(
```

File: test/unit.test.js [hunk 38]
```diff
@@ -3,69 +3,55 @@
...
+      dal.Package.deserialize(
+        { results: dal.Package.fromPublish({ payload: secondFixture.before }) }
```
Comment (jcrugzz) [inline test/unit.test.js]:
```suggestion
        dal.Package.fromPublish(secondFixture.before)
```
